src/pytest_workflow/workflow.py

```diff
@@ -17,54 +17,171 @@
 
     def stderr_to_file(self) -> Path:
-        return bytes_to_file(self.stdout, self.cwd / Path("log.err"))
+        self.wait()
+        return bytes_to_file(self.stderr, self.cwd / Path("log.err"))
+
+    def wait(self):
+        """Waits for the workflow to complete"""
+        # Lock the wait step. Only one waiter is allowed here to wait for
+        # the workflow to complete and write the stderr.
+        # A popen.stderr is a buffered reader and can only be read once.
+        # Once self._stderr and self._stdout are written the lock can be
+        # released
+        with self.wait_lock:
+            while self._popen is None:
+                # This piece of code checks if a workflow has started yet. If
+                # it has not, it waits. A counter is implemented here because:
+                # 1. Incrementing a counter is much faster than checking system
+                #    time
+                # 2. The counter is linked to the self object. This means we
+                #    wait only once for the workflow to start. All consecutive
+                #    wait commands will fail instantly. Having all of these
+                #    wait as well would be a waste of time.
+                if self.wait_timeout_secs is not None and self.wait_counter > (
+                        self.wait_timeout_secs / self.wait_interval_secs):
+                    raise ValueError(
+                        "Waiting on a workflow that has not started within the"
+                        " last {0} seconds".format(self.wait_timeout_secs))
+                time.sleep(self.wait_interval_secs)
+                self.wait_counter += 1
+
+            # Wait for process to finish with _popen.communicate().
+            # _popen.wait() will block with stdout=pipe and stderr=pipe
+            if (self._popen.returncode is None and
+                    self._stderr is None and
+                    self._stdout is None):
+                self._stdout, self._stderr = self._popen.communicate()
 
     @property
     def stdout(self) -> bytes:
-        return self._proc_out.stdout  # for testing log
+        self.wait()
+        return self._stdout  # for testing log
 
     @property
     def stderr(self) -> bytes:
-        return self._proc_out.stderr  # for testing log
+        self.wait()
+        return self._stderr  # for testing log
 
     @property
     def exit_code(self) -> int:
-        return self._proc_out.returncode
+        self.wait()
+        return self._popen.returncode
+
+    def stdout_lines(self):
+        self.wait()
+        return self._stdout.decode().splitlines()
+
+    def stderr_lines(self):
+        self.wait()
+        return self._stderr.decode().splitlines()
 
 
 def bytes_to_file(bytestring: bytes, output_file: Path) -> Path:
     with output_file.open('wb') as file_handler:
         file_handler.write(bytestring)
     return output_file
+
+
+class WorkflowQueue(queue.Queue):
```

Comment:
It might be preferable to have this in a separate file.

Reply:
The file is less than 200 lines including license. What would be the benefits of splitting it up?

Reply:
Agreed with @rhpvorderman here.